File: Lab_5/src/main/scala/lab5/eval/CombinedTerminationCondition.scala
package lab5.eval

import org.uncommons.watchmaker.framework.{PopulationData, TerminationCondition}

class CombinedTerminationCondition(
  combineType: CombinedTerminationCondition.CombineType,
  terminationCondition: TerminationCondition*
) extends TerminationCondition:
  override def shouldTerminate(populationData: PopulationData[_]): Boolean = 
    combineType match
      case CombinedTerminationCondition.CombineType.OR =>
        terminationCondition.exists(_.shouldTerminate(populationData))
      case CombinedTerminationCondition.CombineType.AND =>
        terminationCondition.forall(_.shouldTerminate(populationData))

object CombinedTerminationCondition:
  enum CombineType:
    case AND, OR

File: Additional_1/src/main/java/additional1/eval/CombinedTerminationCondition.java
package additional1.eval;

import org.uncommons.watchmaker.framework.PopulationData;
import org.uncommons.watchmaker.framework.TerminationCondition;

import java.util.Arrays;
import java.util.List;

public class CombinedTerminationCondition implements TerminationCondition {

    public enum CombineType {
        AND,
        OR
    }

    final List<TerminationCondition> conditionList;
    final CombineType combineType;

    public CombinedTerminationCondition(CombineType combineType, TerminationCondition... conditionList){
        this.conditionList = Arrays.stream(conditionList).toList();
        this.combineType = combineType;
    }

    long countTerminated(PopulationData<?> populationData) {
        return conditionList.stream().filter(cond -> cond.shouldTerminate(populationData)).count();
    }

    @Override
    public boolean shouldTerminate(PopulationData<?> populationData) {
        switch (combineType) {
            case OR -> {
                return countTerminated(populationData) > 0;
            }
            case AND -> {
                return countTerminated(populationData) == conditionList.size();
            }
        }
        return false;
    }
}
